src/measure_theory/lp_space.lean

```diff
@@ -0,0 +1,147 @@
+/-
+Copyright (c) 2020 Rémy Degenne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Rémy Degenne.
+-/
+import measure_theory.l1_space
+import analysis.special_functions.pow
+
+/-!
+# ℒp space
+
+This file describes the type of measurable functions with finite seminorm
+`(∫⁻ a, (nnnorm (f a))^(p : ℝ) ∂μ) ^ (1/p)` for `p:ℝ` with `1 ≤ p`.
+
+## Main definitions and results
+
+* `ℒp α β hp1 μ` : the type of measurable functions `α → β` with finite p-seminorm for
```

Comment:
My goal in implementing this is mostly to use it in probability theory. While a large part of that theory can see random variables as their equivalence class, it is not the case in many aspects of the study of stochastic processes with uncountable index sets, hence I wanted to keep the possibility of using a non-quotient type. Of course, defining then the other Lp and showing that it is a normed space is then immediate.
What I wanted to show about Lp is : premetric_space, complete_space (complete uniform space), vector_space. Premetric space is sadly quite weak right now in mathlib (it is a file with less than 10 lemmas and almost no relation to other types), but many properties implemented for metric spaces could be added to premetric (for example the fact that it is also an uniform space).

But I understand that your question is on the use of a type vs simply using the property in_Lp (renamed to mem_Lp). I have still a very limited understanding of why one would want to use a type or not in general, so I don't really know how to answer that.

Reply:
Completeness of premetric spaces is a very weird notion, because there is no uniqueness of the limit -- it can be formulated as: any Cauchy sequence has a limit (but maybe several!). This is too weak to be useful, and that's why there is essentially nothing around this in mathlib (I only introduced it to be able to take a quotient right away and obtain the corresponding metric space).

The question of working with a property or introducing a subtype is a difficult one, for which there is no definitive answer (it is the unbundled/bundled discussion that shows up often on Zulip): it really depends on the context, the applications, the pros and cons of both approaches (and we got it wrong several times, and learnt the hard way that group morphisms should be bundled, for instance). Here, I don't see a lot of gain, and I anticipate some pain (for instance, saying that the product of functions in `L^p` and `L^q` is in `L^r` when `1/r = 1/p + 1/q` is straightfoward with the unbundled approach, but it is more painful with the bundled (subtype) approach) and a lot of code duplication. My gut feeling is that using the unbundled approach at the level of functions (i.e., use properties) and the bundled approach at the level of equivalence classes of functions (define a subtype, to have a metric space) would be the most efficient, but I might very well be wrong on this!

Reply:
looking at the code I have (beyond that PR), I realize that it is already a sequence of statements about the property mem_Lp, followed in the end by a few one-line lemmas to translate the results to the type. I'll drop the type for now.